### PR TITLE
#3181

### PR DESCRIPTION
**Pull request by vestrada**
https://github.com/craftercms/craftercms/issues/3181 - [studio] Support user controlled publishing of soft and hard dependencies #3181

### CODE REVIEW

File: static-assets/components/cstudio-dialogs-templates/approve.html [hunk 1]
```diff
@@ -20,7 +20,8 @@
     <h3 class="view-title" data-translation="approveDialogTitle">Approve for Publish</h3>
 
     <div class="view-caption" data-translation="approveDialogSubTitle">
-        The following checked items will go live.
+        Selected files will go live upon submission. Hard dependencies are automatically submitted with the main items.
```
Comment (rart) [inline static-assets/components/cstudio-dialogs-templates/approve.html]: Use: "Selected files will be published. Hard dependencies are automatically included. Soft dependencies are optional and you may choose which to include."

Add this to a message bundle and translate accordingly.

File: ui/app/src/components/DependencyItem.tsx [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
+import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
+import Typography from '@material-ui/core/Typography';
+import Checkbox from '@material-ui/core/Checkbox';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+import { get } from '../utils/ajax';
+import { Item } from '../models/item';
+
+// components/DependecyItem.tsx
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: Remove comment

File: ui/app/src/components/DependencyItem.tsx [hunk 3]
```diff
@@ -0,0 +1,206 @@
+import React, { useEffect, useState } from 'react';
+import { withStyles } from '@material-ui/core/styles';
+import MuiExpansionPanel from '@material-ui/core/ExpansionPanel';
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: Expansion panel is no longer in use. Please remove all references to it including unused styles for it.

File: ui/app/src/components/DependencyItem.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
+import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
+import Typography from '@material-ui/core/Typography';
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: Remove the usage of Typography component

File: ui/app/src/components/DependencyItem.tsx [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                        ))
+                      ) 
+                      : (
+                        <div></div>
+                      )
+                    }
+                  </ul>
+                    <p>
+                      <strong>Soft Dependencies</strong> • <span>Submission Optional</span> 
+                      <button className="navButtons" onClick={selectAllSoft}>Select All</button>
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: Dasherized classnames. This class is probably not even used so you should remove it completely if that's the case.

File: ui/app/src/components/DependencyItem.tsx [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                  <ul>
+                    {
+                      deps && deps.hard
+                      ? ( 
+                        deps.hard.map((uri: string) => (
+                          <li key={uri}>{uri}</li>
+                        ))
+                      ) 
+                      : (
+                        <div></div>
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: replace with `null`

File: ui/app/src/components/DependencyItem.tsx [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                              onChange={(e) => props.onSelection(uri, e.target.checked)}
+                              value={uri}
+                              color="primary"
+                            />
+                            {uri}
+                          </li>
+                        ))
+                      ) 
+                      : (
+                        <div></div>
```
Comment (rart) [inline ui/app/src/components/DependencyItem.tsx]: replace with `null`

File: ui/app/src/models/item.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+/*
```
Comment (rart) [inline ui/app/src/models/item.ts]: Model name (file name) should be upper case

File: ui/scss/_studio-view.scss [hunk 9]
```diff
@@ -481,3 +489,127 @@ table.scroll-body {
...
+
+.submission-comment{
+  height: 111px!important;
+}
+
+.dependencies-display{
+  height: 444px;
+}
+
+//Publish - soft hard dependencies.
```
Comment (rart) [inline ui/scss/_studio-view.scss]: Slightly unorthodox but do check if you can import "../route/to/the/file/on/the/app/directory"

File: ui/app/src/components/DependecySelection.tsx [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                                value={uri}
+                                color="primary"
+                              />
+                            </div>
+                            <div className="dependency-selection--list--soft-item" >{uri}</div>
+                          </li>
+                        ))
+                      )
+                      : (
+                        <div></div>
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: replace with `null`

File: ui/app/src/components/DependecySelection.tsx [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+                  Show All Dependencies
+                </button>
+              ) : (
+                  <div></div>
+                )
+            )
+        }
+        <p>
+          Changes in the selection of items to publish will require "all dependencies" to be recalculated.
+          Dependency calculation is computationally.
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: Remove "Dependency calculation is computationally."

File: ui/app/src/components/DependecySelection.tsx [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                <ul className="dependency-selection--list">
+                  {
+                    deps && deps.hard
+                      ? (
+                        deps.hard.map((uri: string) => (
+                          <li className="dependency-selection--list--hard" key={uri}>{uri}</li>
+                        ))
+                      )
+                      : (
+                        <div></div>
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: replace with `null`

File: ui/app/src/components/DependecySelection.tsx [hunk 13]
```diff
@@ -0,0 +1,232 @@
+import React, { useState } from 'react';
+import { Item } from '../models/item';
+import '../styles/dependency-selection.scss';
+import { withStyles } from '@material-ui/core/styles';
+import Checkbox from '@material-ui/core/Checkbox';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import { get } from '../utils/ajax';
+
+// components/DependecySelection.tsx
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: Remove comment

**Review by rart — CHANGES_REQUESTED** [review]
- Multiple translations missing (don't see message bundle file(s) updated on this PR)
- App.tsx should not be committed. Roll it back and don't include in PR.
- Please also remove `DependencyItem.tsx` from the PR. You may ignore all comments around it.

File: ui/app/src/components/DependecySelection.tsx (hunk 13 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: Done

File: ui/app/src/components/DependecySelection.tsx (hunk 12 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: Done

File: ui/app/src/components/DependecySelection.tsx (hunk 11 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: Done

File: ui/app/src/components/DependecySelection.tsx (hunk 10 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: Done

File: ui/scss/_studio-view.scss (hunk 9 above)
Comment (vestrada) [reply ui/scss/_studio-view.scss]: Done

File: ui/app/src/components/DependencyItem.tsx (hunk 5 above)
Comment (vestrada) [reply ui/app/src/components/DependencyItem.tsx]: file is removed

File: ui/app/src/components/DependencyItem.tsx (hunk 4 above)
Comment (vestrada) [reply ui/app/src/components/DependencyItem.tsx]: file is removed

File: ui/app/src/components/DependencyItem.tsx (hunk 3 above)
Comment (vestrada) [reply ui/app/src/components/DependencyItem.tsx]: file is removed

File: ui/app/src/components/DependencyItem.tsx (hunk 2 above)
Comment (vestrada) [reply ui/app/src/components/DependencyItem.tsx]: file is removed

File: static-assets/components/cstudio-dialogs-templates/approve.html (hunk 1 above)
Comment (vestrada) [reply static-assets/components/cstudio-dialogs-templates/approve.html]: Done

File: ui/app/src/components/DependecySelection.tsx [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+      {}
+    )
+  );
+  const setChecked = (uri: string[], isChecked: boolean) => {
+    const nextChecked = { ...checked };
+    (Array.isArray(uri) ? uri : [uri]).forEach((u) => {
+      nextChecked[u] = isChecked;
+    });
+    setRef();
+    _setChecked(nextChecked)
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: `;` missing

File: ui/app/src/components/DependecySelection.tsx [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  }
+
+  const [checkedSoftDep, _setCheckedSoftDep] = useState<any>({});
+  const setCheckedSoftDep = (uri: string[], isChecked: boolean) => {
+    const nextCheckedSoftDep = { ...checkedSoftDep };
+    (Array.isArray(uri) ? uri : [uri]).forEach((u) => {
+      nextCheckedSoftDep[u] = isChecked;
+    });
+    setRef();
+    _setCheckedSoftDep(nextCheckedSoftDep)
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: ;

File: ui/app/src/components/DependecySelection.tsx [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+  }
+  const cleanCheckedSoftDep = () => {
+    const nextCheckedSoftDep = {};
+    setRef();
+    _setCheckedSoftDep(nextCheckedSoftDep)
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: ;

File: ui/app/src/components/DependecySelection.tsx [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+
+  function setRef() {
+    result.current = Object.entries({ ...checked, ...checkedSoftDep })
+      .filter(([key, value]) => value === true)
+      .map(([key]) => key)
+  }
+
+  function showAllDependencies() {
+    setshowDepBtn(false);
+    get(`/studio/api/2/dependency/dependencies?siteId=${siteId}&paths=${Object.entries({ ...checked })
```
Comment (rart) [inline ui/app/src/components/DependecySelection.tsx]: *Object.entries()***.join(',')**

File: ui/app/src/models/item.ts (hunk 8 above)
Comment (rart) [reply ui/app/src/models/item.ts]: Pending change.

**Review by rart — CHANGES_REQUESTED** [review]
**static-assets/components/cstudio-view-controllers/approve.js**
- Remove the functions that were attached to the `.select-all-check` and `.show-all-deps` buttons (i.e. showAllDepsActionClicked and showAllDepsActionClicked)
- There seem to be 2 startup props, meaning one of them does nothing. Please clean up.

**request-publish.js**
- Clean up old action handler for `.show-all-deps`
- Remove commented code

See other comments

File: ui/app/src/components/DependecySelection.tsx (hunk 17 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: Per our conversation, this was ok I just created new const.

File: ui/app/src/components/DependecySelection.tsx (hunk 16 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: done

File: ui/app/src/components/DependecySelection.tsx (hunk 15 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: done

File: ui/app/src/components/DependecySelection.tsx (hunk 14 above)
Comment (vestrada) [reply ui/app/src/components/DependecySelection.tsx]: done

File: ui/app/src/models/item.ts (hunk 8 above)
Comment (vestrada) [reply ui/app/src/models/item.ts]: done

File: static-assets/components/cstudio-common/resources/de/base.js [hunk 18]
```diff
@@ -451,6 +451,16 @@ CStudioAuthoring.Messages.registerBundle("forms", "de", {
     requestApproveDialogSubTitle: "Wann sollen die ausgewählten Elemente publiziert werden?",
     requestApproveDialogEmailMe: "Benachrichtung per E-Mail erhalten, wenn die Elemente zum Publizieren freigegeben wurden.",
 
+    itemsForPublish:"Items to Publish",
```
Comment (rart) [inline static-assets/components/cstudio-common/resources/de/base.js]: Translate

File: static-assets/components/cstudio-common/resources/kr/base.js [hunk 19]
```diff
@@ -422,6 +422,16 @@ CStudioAuthoring.Messages.registerBundle("forms", "kr", {
     requestApproveDialogSubTitle: "당신이 원하는 때 체크 항목 (들)을 게재하는 방법?",
     requestApproveDialogEmailMe: "항목이 게시가 승인되면 이메일을 보내십시오.",
 
+    itemsForPublish:"Items to Publish",
```
Comment (rart) [inline static-assets/components/cstudio-common/resources/kr/base.js]: Translate

**Review by rart — APPROVED** [review]

**Comment by rart** [comment]
Message localization updated on https://github.com/craftercms/studio-ui/pull/1307

**Pull request merged**